### PR TITLE
01/2023 Coverity fixes

### PR DESCRIPTION
**Pull request by szalpal**
Signed-off-by: szalpal <mszolucha@nvidia.com>

## Category:

**Bug fix** (*non-breaking change which fixes an issue*)

## Description:

Coverity fixes for January 2023.

### Tests:

- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

File: dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc [hunk 1]
```diff
@@ -89,7 +97,7 @@ NvJpeg2000DecoderInstance::~NvJpeg2000DecoderInstance() {
       res.intermediate_buffer.free();
     });
 
-  for (auto thread_id : tp_->GetThreadIds())
+  for (auto& thread_id : tp_->GetThreadIds())
```
Comment (jantonguirao) [inline dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc:100]: https://en.cppreference.com/w/cpp/thread/thread/id
```The class thread::id is a lightweight, trivially copyable class that serves as a unique identifier of std::thread objects.```
I think we are good with copy here

File: dali/operators/reader/loader/video/frames_decoder_gpu.h [hunk 2]
```diff
@@ -78,7 +78,7 @@ class NVDECLease {
     }
 
     ~NVDECLease() {
-      Return();
+      Return();  // Intentionally throwing in noexcept scope.
```
Comment (jantonguirao) [inline dali/operators/reader/loader/video/frames_decoder_gpu.h:81]: nitpick: This comment here is a bit misleading (I don't see that we are throwing here). Also, I think that we have this widely accepted (allow exceptions to arise on dtors) that we don't need to add those comments

File: dali/test/cfitsio_test.cc [hunk 3]
```diff
@@ -27,13 +27,17 @@ class CfitsioTest : public ::testing::Test {
   char filename[40];
   fitsfile *fptr;
 
-  virtual void SetUp() {
+  void SetUp() final {
     status = 0;
-    remove("testprog.fit");
+    if (remove("testprog.fit") != 0) {
```
Comment (jantonguirao) [inline dali/test/cfitsio_test.cc:32]: to be on the safe side, this file should proably live in a temporary dir

File: dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc (hunk 1 above)
Comment (mzient) [reply dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc:100]: :+1:

File: dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc [hunk 4]
```diff
@@ -58,10 +58,18 @@ NvJpeg2000DecoderInstance::NvJpeg2000DecoderInstance(
...
+  bool resource_allocation_successful = true;
   ForEachThread(*tp_, [&](int tid) noexcept {
-    CUDA_CALL(cudaSetDevice(device_id));
-    per_thread_resources_[tid] = {nvjpeg2k_handle_, device_memory_padding, device_id_};
+    try {
+      CUDA_CALL(cudaSetDevice(device_id));
+      per_thread_resources_[tid] = {nvjpeg2k_handle_, device_memory_padding, device_id_};
+    } catch (dali::CUDAError&) {
+      resource_allocation_successful = false;
+    }
```
Comment (mzient) [inline dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc:68]: This is just wrong. If you want to fix the problem, not just silence the tool, then it should be fixed at `ForEachThread`, not here (here you could just remove the noexcept).
Also, this piece of code could still throw a plain std::bad_alloc.

File: dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc [hunk 5]
```diff
@@ -58,10 +58,18 @@ NvJpeg2000DecoderInstance::NvJpeg2000DecoderInstance(
...
   });
+  if (!resource_allocation_successful) {
+    throw std::runtime_error("Unsuccessful resource allocation for NvJpeg2000 decoder.");
+  }
```
Comment (mzient) [inline dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc:72]: Please remove.

File: dali/operators/reader/loader/video/frames_decoder_gpu.h (hunk 2 above)
Comment (mzient) [reply dali/operators/reader/loader/video/frames_decoder_gpu.h:81]: :+1: This comment belongs in coverity report, not in the code.

File: dali/pipeline/operator/builtin/input_operator.h [hunk 6]
```diff
@@ -127,7 +127,7 @@ class InputOperator : public Operator<Backend>, virtual public BatchSizeProvider
 
 
   virtual ~InputOperator() {
-    sync_worker_.ForceStop();
+    sync_worker_.ForceStop();  // Intentionally throwing in noexcept scope.
```
Comment (mzient) [inline dali/pipeline/operator/builtin/input_operator.h:130]: Likewise.

File: dali/test/cfitsio_test.cc (hunk 3 above)
Comment (mzient) [reply dali/test/cfitsio_test.cc:32]: True, but out of scope.

File: include/dali/core/util.h [hunk 7]
```diff
@@ -87,21 +88,25 @@ constexpr Value align_up(Value v, Alignment a) {
 
 DALI_HOST_DEV
 constexpr int32_t div_ceil(int32_t total, uint32_t grain) {
+  assert(grain != 0);
```
Comment (mzient) [inline include/dali/core/util.h:91]: No, no, no and no.

**Review by mzient — CHANGES_REQUESTED** [review]
Multiple issues. Don't try too hard to please the tool.
1. As a general rule, don't put asserts or exception handling in lightweight math. `sqrt` or plain division don't throw or abort and so shouldn't they.
2. Exception handling in noexcept functions - either do it properly or let it terminate. It's better than swallowing the exception and producing a less informative one (which is unlikely to be handled anyway).

**Comment by szalpal** [comment]
Closing the PR due to disagreement with one of the reviewers.

**Pull request closed without merging**